synapse/handlers/federation.py

```diff
@@ -963,14 +963,17 @@ async def get_state_ids_for_pdu(self, room_id: str, event_id: str) -> List[str]:
 
         state_map = next(iter(state_groups.values()))
 
-        if event.is_state():
-            # Get previous state
+        # we need the state *before* event.
+        state_key = event.get_state_key()
+        if (
+            state_key is not None
+            and state_map.get((event.type, state_key)) == event.event_id
```

Comment:
I'm struggling to see why this extra check against the event ID; could it be that sometimes `state_map` might already have the previous state?

Reply:
hrm, good question.

`state_map` certainly won't include `event` if `event` was rejected... but it's not clear to me what the behaviour of this method is when `event` is rejected anyway.

I guess I'll have to poke at this a bit more to understand it.

Reply:
turns out that we never get this far if the event was rejected (i'll put a sytest up for that in a bit), so I really can't think of any reason `state_map` wouldn't include `event`. Accordingly, I've replaced this with an `assert`.

Reply:
Thanks for investigating! The comments are clear now!